Adding STS to the aws backend

dgromov opened the pull request:
Added an /sts endpoint to the aws backend that can pull down sts credentials. The logic is very similar to the existing iam credentials logic but the keys provided are ready to use immediately. 

Related: 
https://github.com/hashicorp/vault/issues/687
https://github.com/hashicorp/vault/issues/295

jefferai commented:
Thanks for contributing this! I'm trying to corral some reviewers over.

dgromov commented:
@jefferai -- What is the policy for adding the github.com/aws/aws-sdk-go/service/sts dependency? ```godep save``` changes a lot of files that aren't related to the change here.

jefferai commented:
@dgromov It's no problem to add it, but doing it in a way that doesn't mess things up is the trick.

My general method for adding to `godeps` is:

1. `godep restore`
2. `rm -rf Godeps`
3. <ensure whatever new library/update you need is present in your GOPATH>
4. `godep save ./...`
5. Examine result, ensure only expected changes occur

jantman commented:
Note that I have virtually no Go experience at all (like... took me a couple of tries to get this to build right), so I'm not looking at the code itself, just the functionality.

* As far as I can tell, this works. With the [testing that I did](https://gist.github.com/jantman/1d50ca7c8b3be0db8da8), the current IAM auth requires ~2s wait for EC2 queries to work, but works immediately with STS.
* Can you update the documentation for this? I was hoping to see some, but managed to muddle my way through the code enough to figure out how it should work.
* My only comment is that about the logging output; see the comment on that specific line.

jantman commented on builtin/logical/aws/secret_access_keys.go:
```diff
@@ -37,6 +42,75 @@ func secretAccessKeys(b *backend) *framework.Secret {
+	}
+
+	// Create the keys and token
+	fmt.Println(username, policy)
```
Is there a proper way to handle this logging? It looks quite ugly in the output of ``vault server -dev -log-level=trace``:

```
2016/01/14 08:06:12 [DEBUG] rollback: attempting rollback on secret/
2016/01/14 08:06:12 [DEBUG] rollback: attempting rollback on aws/
2016/01/14 08:06:12 [DEBUG] rollback: attempting rollback on cubbyhole/
2016/01/14 08:06:12 [DEBUG] rollback: attempting rollback on sys/
vault-root-1452776776-3632 {"Version":"2012-10-17","Statement":{"Effect":"Allow","Action":"*","Resource":"*"}}
2016/01/14 08:07:12 [DEBUG] rollback: attempting rollback on aws/
2016/01/14 08:07:12 [DEBUG] rollback: attempting rollback on secret/
2016/01/14 08:07:12 [DEBUG] rollback: attempting rollback on cubbyhole/
```

I don't know if Vault has a real logging subsystem, or if there's just a better way to format this with the date, and ideally with some indication of where it's coming from?

jefferai replied:
Logging is available to all backends, so it's just a matter of this one using the proper one. @dgromov see https://godoc.org/github.com/hashicorp/vault/logical/framework#Backend.Logger

dgromov commented:
@jantman The print statement you were talking about was a debug line I left in by accident. I took it out. I also updated the docs at website/source/docs/secrets/aws/index.html.md. Thank you for that gist btw. 

@jefferai The duration field I added to the /sts path has a sane(I think) default of an hour but is not configurable since that path is a read operation. Do you have any thoughts on how to make it configurable? I've though of adding it to the path like: /sts/[role_name]/[duration] but am unclear on how to make it optional in that case.

jefferai commented:
@dgromov Your diff has gotten crazy -- maybe you need a rebase? So I can't easily see your changes here, but:

The AWS backend has very bad bugs w.r.t. lease handling as it is, particularly one where it checks the configured lease duration for a renew operation, but not for an initial issuance of creds(!).

That said -- it seems to me like the right thing to do is honor the existing configurable lease endpoint, and fix the behavior so it actually uses that for initial credential issuance.

In not the next release, but hopefully one of the couple after, there will be some big changes in the framework underlying these backends that will make handling roles and leases far more sane.

dgromov commented:
@jefferai the huge change set is because of the new deps. Following the instructions you gave above adds "golang.org/x/sys/unix" to the Godeps/Godeps.json file. I'm not entirely sure why though.

```
               {
+                       "ImportPath": "github.com/aws/aws-sdk-go/service/sts",
+                       "Comment": "v1.0.8-3-g64ecfaa",
+                       "Rev": "64ecfaae861c2f24bb2e8cf47c257a40e2b4f0d4"
+               },
+               {
                        "ImportPath": "github.com/bgentry/speakeasy",
                        "Rev": "36e9cfdd690967f4f690c6edcc9ffacd006014a0"
                },
@@ -287,6 +292,10 @@
                        "Rev": "2baa8a1b9338cf13d9eeb27696d761155fa480be"
                },
                {
+                       "ImportPath": "golang.org/x/sys/unix",
+                       "Rev": "442cd600860ce722f6615730eb008a37a87b13ee"
+               },
+               {
                        "ImportPath": "gopkg.in/asn1-ber.v1",
                        "Comment": "v1.1",
                        "Rev": "4e86f4367175e39f69d9358a5f17b4dda270378d"
```

jefferai commented:
@dgromov I'm not sure -- it might be due to crypto needs in STS.

Can you back that change out? I can update the Godeps once it's merged in, but let's get the diff happy again.

dgromov commented:
@jefferai Ok, backed out the dependency change. Diff seems reasonable now. I figured you'd want the CI tests passing before you merged :) 

W.r.t the lease. The STS duration is a temporary session that expires on its own after a while. It is described here: http://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_request.html. I've set it up to expire in 1 hour which matches the existing duration default that gets assigned to all new creds. Would it be cool to revisit making that nicer once those backend changes are done?

jefferai commented:
Given that the default and maximum for STS is 1 hour, and that's already the default for AWS (since it doesn't honor TTL except for renewing(!)), I'm fine with it being hardcoded to 1 hour for now. As you said, this can be revisited later.

dgromov commented:
Great, thanks @jefferai. If @jantman feels I've adequately addressed his issues, I'm ready for this to be merged.

jefferai commented:
Hi @dgromov,

Just did a code review. Looks good, with one question: weren't you going to hardcode a 1-hour value? If not (which is fine!), then:

1. Please change the "duration" value to "ttl" to be more (theoretically) consistent with other terminology in Vault. It's inconsistent across backends but we're slowly trying to fix that. (Don't forget the documentation!  :-)  )
2. Use a `framework.TypeDurationSecond` type for the value. This is (again not consistently used :-( ) but our preferred method of specifying TTLs. It's similar to using a string and calling `time.ParseDuration`, but it does that (and associated validation) automatically for you. So when you actually want the value, you type-assert it as `.(int)`, not `.(string)` and you get an integer number of seconds. This lets people pass in either an integer number of seconds if they like (e.g. `3600`) but also allows them to pass in a string format, e.g. `1h`.

Other than that, looks good!

jantman commented:
Sorry for the delays, @dgromov and @jefferai. This looks good to me with one exception: can we please add a small bit somewhere to the docs on STS that this:

1. creates an IAM user with a random name
2. uses STS' [GetFederationToken](http://docs.aws.amazon.com/STS/latest/APIReference/API_GetFederationToken.html) API call

This might help others, especially given:

* Some people - like my employer - are so close to their IAM user limits, that if we were using Vault and STS in a bunch of Jenkins jobs, with 1 hour leases we'd probably run out, and,
* I think most people are really only aware of STS' AssumeRole API calls, which let you assume a role in another account. I imagine that without clarification, many people would read "STS" and equate it with "cross-account access"

Thanks!

jefferai commented:
Just curious, because I don't know, is there some other method of acquiring
STS tokens than how it's done here, with creating an IAM user? If so, I
wonder if that would be a better approach?

jantman commented:
@jefferai good point.

Per the [GetFederationToken](http://docs.aws.amazon.com/STS/latest/APIReference/API_GetFederationToken.html) documentation,

> The passed policy is attached to the temporary security credentials that result from the GetFederationToken API call--that is, to the federated user. When the federated user makes an AWS request, AWS evaluates the policy attached to the federated user in combination with the policy or policies attached to the IAM user whose credentials were used to call GetFederationToken. AWS allows the federated user's request only when both the federated user and the IAM user are explicitly allowed to perform the requested action. The passed policy cannot grant more permissions than those that are defined in the IAM user policy.

I hadn't really considered this until now. By my reading, this means that the IAM User that Vault auths with (those written into ``aws/config``) must have a superset of the permissions granted to the federated user; i.e. just allowing Vault to create IAM resources is not enough. Assuming that's the case (I can test next week), that _really_ needs to be documented clearly.

Also assuming that's the case, I don't think we need to create IAM users. I think we can just pass the policy in to the ``GetFederationToken`` and make that call as the same user that Vault auths with.

dgromov commented:
@jefferai @jantman -- Good catch on the IAM User. I went back through the code an realized that the generated user wasn't being used for anything. It now uses the aws/config creds to generate the token and that's it. A side-effect of that is that there is no rollback required for the STS approach. AWS will clean those up on its own. (As far as I understand)

@jantman You are right about the superset of privileges thing. I just confirmed it and added that to the aws docs. Please let me know if that can/should be clearer.

@jefferai I changed the type of ```ttl``` to be ```framework.TypeDurationSecond```. You are correct that I don't formally call it a constant, but ```read aws/sts ``` has no way to set it. So, its effectively a constant... Not pretty I guess... I'm open to suggestions.

jantman commented:
I haven't finished my morning coffee yet, so I might've missed something, but this looks good to me. Thanks so much!

dgromov commented:
Thanks. I also added another line to the doc about the 1hr ttl. Figured that would be a nice detail :)

dgromov commented:
@jefferai, just checking in. Is there anything else you'd like me to do here?

jefferai commented:
@dgromov Just got back from vacation, will look at the changes soon!

jefferai commented:
@dgromov Any thoughts on https://github.com/hashicorp/vault/issues/295#issuecomment-161482446 ?

jefferai commented on builtin/logical/aws/secret_access_keys.go:
```diff
@@ -123,6 +173,16 @@ func (b *backend) secretAccessKeysRenew(
 	req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
 	// Get the username from the internal data
+	isSTS, ok := req.Secret.InternalData["is_sts"]
+	if !ok {
```
This will break upgrades -- existing secrets won't be able to revoke. This needs to gracefully handle the lack of `is_sts` by assuming IAM (e.g. just forcing `isSTS` to `false`)

jefferai replied:
(See comment below)

jefferai commented:
Seems good to me besides my one comment above and your thoughts on the comment from the other issue.

jefferai commented:
Ah, one other thing: I just merged in a PR that makes some decent changes to AWS. I was going to ask you to take a look at the changes and which ones should be incorporated into secretAccessKeysAndTokensCreate -- but I realized that there is so much copied code, I wonder if it makes more sense to instead simply have secretAccessKeysCreate take a bool flag indicating whether STS should be used, and use that bool to enable or disable appropriate blocks of code. That way the shared things (such as the changed username format) get updated consistently for both code paths.

jefferai commented on builtin/logical/aws/secret_access_keys.go:
```diff
@@ -123,6 +173,16 @@ func (b *backend) secretAccessKeysRenew(
+	}
+
+	// STS cleans up after itself
+	if isSTS.(bool) {
```
This can panic if, for some reason, it's not a bool. I think the best thing to do would be instantiate a bool var (with it defaulting to nil), then use the `isSTS, ok := req.Secret.InternalData["is_sts"]` call to drop into this check if `ok` is true. Use the dual-value version here. So something like (not tested/checked):
```
stsInt, ok := req.Secret.InternalData["is_sts"]
if ok {
  isSTS, ok = stsInt.(bool)
  if ok {
    if isSTS {
      return nil, nil
    } // else, it was set, but not to true, so treat it as IAM by falling through
  } else {
    return nil, fmt.Errorf("secret has is_sts but value could not be understood")
  }
}

dgromov commented:
@jefferai w.r.t to #295, if we're using the root Vault user to make the STS creds, we can't just delete that user. Also, the STS documentation directly says that one cannot revoke the generated keys. However, there is a way to explicitly deny access to resources that were already made. See: http://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_control-access_disable-perms.html

Until AWS has a way to revoke STS creds, this is just a limitation of this approach. People still have the option of using the existing IAM mechanism if they are concerned about being able to revoke things as needed.

jefferai commented:
@dgromov Fair enough, I guess that comment is moot since you're no longer creating IAM users to then use for STS issuance.

dgromov commented:
To your other point, I agree about both of the bools. Will refactor a bit.

dgromov commented:
@jefferai 

- I added the check for is_sts that you mentioned. It is pretty much the snippet you pasted.
- It turns out the secretAccessKeysCreate and secretAccessKeysAndTokenCreate share next to nothing now outside of the username. I factored the username into its own function so they can share that. 
- I think your PR merge introduced a bug into the acceptance test. There was a call to WriteOperation instead of UpdateOperation that was fixed from a commit 2 weeks ago. I changed it back to UpdateOperation.

jefferai commented:
Awesome, thanks. The normal unit tests do not run acceptance tests (one day, one day) so it's easy to miss things like that. I'll give another review and merge if it all looks good!

jefferai commented:
LGTM. Merging!
